### PR TITLE
stable dev

### PR DESCRIPTION
**Pull request by desaibhargav**
This PR is the "stable" version of `nett-benchmarks`, in the sense that it:

1. Makes a couple of bug fixes that broke `dev` even for vanilla use-cases (no jobs start, pipeline breaks at `.run()`)
2. Addresses #109, all jobs complete as expected (even with `save_checkpoints=True`). Running multiple brains with a custom encoder and `save_checkpoints` works as expected.  
3. Address #30, the python multiprocessing pool needs to be manually teared down if calling `ProcessPoolExecutor` without context (not using `with`)
4. Now allows passing either of the `["small", "medium", "large"]` to the `encoder` argument while initializing the `Brain`. The appropriate `encoder` will automatically be picked up. 
For example, 
```
brain = Brain(policy="CnnPolicy", algorithm="PPO", encoder='small')
``` 
5. Fixes a small type in the README, where `dvs` is shown as a valid argument to be passed to `Body` even after removing this behavior

### CODE REVIEW

File: src/nett/brain/__init__.py [hunk 1]
```diff
@@ -72,6 +70,12 @@ def get_encoder_dict() -> dict[str, str]:
             encoder_class = [node for node in ast.walk(module) if isinstance(node, ast.ClassDef)][0]
             # add to the dictionary
             encoders_dict[module_name] = encoder_class.name
+
+    # add small, medium and large encoders too
+    encoders_dict['small'] = 'defualt'
```
Comment (Zach-Attach) [inline src/nett/brain/__init__.py]: Typo
```suggestion
    encoders_dict['small'] = 'default'
```

File: src/nett/brain/builder.py [hunk 2]
```diff
@@ -322,10 +323,14 @@ def _validate_encoder(encoder: Any | str) -> BaseFeaturesExtractor:
...
         if isinstance(encoder, str):
-            if encoder not in encoder_dict.keys():
-                raise ValueError(f"If a string, should be one of: {encoder_dict.keys()}")
-            encoder = getattr(encoders, encoder_dict[encoder])
-
+            if encoder in encoders_dict.keys() and encoder != 'small':
+                encoder = getattr(encoders, encoders_dict[encoder])
+            elif encoder == 'small':
+                encoder = NatureCNN
+                encoder.__name__ = 'NatureCNN'
```
Comment (Zach-Attach) [inline src/nett/brain/builder.py]: Minor simplification / cleanup:
```suggestion
            if encoder == 'small':
                encoder = NatureCNN
                encoder.__name__ = 'NatureCNN'
            elif encoder in encoders_dict.keys():
                encoder = getattr(encoders, encoders_dict[encoder])
```

File: src/nett/nett.py [hunk 3]
```diff
@@ -375,25 +382,22 @@ def _launch_jobs(self, jobs: list[Job], wait: bool, waitlist: list[Job], verbose
...
                     job_future = executor.submit(self._execute_job, job)
                     job_sheet[job_future] = job
                     time.sleep(1)
-
-            if wait:
-                while job_sheet:
-                    done, _ = future_wait(job_sheet, return_when=FIRST_COMPLETED)
-                    for doneFuture in done:
-                        job_sheet.pop(doneFuture)
-                    time.sleep(1)
```
Comment (Zach-Attach) [inline src/nett/nett.py]: Why are we getting rid of the synchronous flag? I would prefer an option to not have the code sit in the background

File: src/nett/nett.py [hunk 4]
```diff
@@ -101,13 +101,13 @@ def run(self,
...
         # calculate iterations
-        iterations: dict[str, int] = {
-            "train": steps_per_episode * train_eps,
-            "test": test_eps * self.environment.num_test_conditions
-        }
-
-        if not issubclass(self.brain.algorithm, RecurrentPPO):
-            iterations["test"] *= steps_per_episode
+        iterations: dict[str, int] = {}
+        if mode in ["train", "full"]:
```
Comment (Zach-Attach) [inline src/nett/nett.py]:
```suggestion
        if mode in ["train", "full"] or job_memory == "auto": #TODO: auto job_memory runs a train job, requiring iterations["train"] to run
```
auto memory estimation currently can only use a "train" job to estimate memory usage (which will always use more memory than a "test" job so it is the safer of the two options). Until auto memory estimation is adjusted to support accurate test job memory usage, I recommend adding this.

File: src/nett/nett.py [hunk 5]
```diff
@@ -286,6 +289,10 @@ def _execute_job(self, job: Job) -> Future:
     def _estimate_job_memory(self, devices: list[int], base_port: int) -> int:
         self.logger.info("Estimating memory for a single job")
         try:
+            # create a temporary directory to hold memory estimate during runtime
+            tmp_path = Path("./.tmp/").resolve()
+            tmp_path.mkdir(parents=True, exist_ok=True)
+
```
Comment (Zach-Attach) [inline src/nett/nett.py]: tmp path resulted in some issues and was removed in recent versions. Now uses brain_0 to store this and deletes it upon completion.
```suggestion
```

File: src/nett/nett.py [hunk 6]
```diff
@@ -363,7 +370,7 @@ def _launch_jobs(self, jobs: list[Job], wait: bool, waitlist: list[Job], verbose
             dict[Future, Job]: A dictionary of futures corresponding to the jobs that were launched from them.
         """
         try:
-            max_workers = 1 if len(jobs) == 1 else None
+            max_workers = 1 if len(jobs) == 1 else os.cpu_count()
```
Comment (Zach-Attach) [inline src/nett/nett.py:371]: Interesting. So the problem was too many workers?

File: src/nett/utils/job.py [hunk 7]
```diff
@@ -75,6 +75,15 @@ def _configure_paths(self) -> dict[str, Path]:
...
     return paths
+  
+  def validation_kwargs(self) -> dict[str, Any]:
+    """Get the environment kwargs for validation
+    """
+    kwargs = self.env_kwargs()
+    kwargs["validation-mode"] = True
+    kwargs["record-chamber"] = False
+    kwargs["record-agent"] = False
+    return kwargs
```
Comment (Zach-Attach) [inline src/nett/utils/job.py:86]: Good catch! My bad

**Review by Zach-Attach — CHANGES_REQUESTED** [review]
Looks like a great addition! A few changes need to be made first, but otherwise good work

File: src/nett/brain/__init__.py (hunk 1 above)
Comment (desaibhargav) [reply src/nett/brain/__init__.py]: Ah! Nice, will fix this, thanks

File: src/nett/nett.py (hunk 3 above)
Comment (desaibhargav) [reply src/nett/nett.py]: I took it out for debugging, I'll add it again

File: src/nett/nett.py (hunk 5 above)
Comment (desaibhargav) [reply src/nett/nett.py]: oh thank god, I really don't like temp paths and temp files

File: src/nett/nett.py (hunk 6 above)
Comment (desaibhargav) [reply src/nett/nett.py:371]: I really do not want to believe this. I'm running a test with `None`, let's see if that works

**Comment by desaibhargav** [comment]
@Zach-Attach I've made the changes, I haven't added the synchro block yet, I wasn't very sure where to put it, would you want to add that after we merge to dev?

File: src/nett/nett.py [hunk 8]
```diff
@@ -351,7 +354,7 @@ def _get_memory_status(devices: list[int]) -> dict[int, dict[str, int]]:
         }
         return memory_status
     
-    def _launch_jobs(self, jobs: list[Job], wait: bool, waitlist: list[Job], verbose: bool) -> dict[Future, Job]:
+    def _launch_jobs(self, jobs: list[Job], waitlist: list[Job], verbose: bool) -> dict[Future, Job]:
```
Comment (Zach-Attach) [inline src/nett/nett.py]: wait will need to be added back (this is how synchronous is passed). If you want to change the name, feel free, but this needs to stay in one way or another

File: src/nett/nett.py [hunk 9]
```diff
@@ -143,7 +143,7 @@ def run(self,
 
         # launch jobs
         self.logger.info("Launching")
-        job_sheet = self._launch_jobs(jobs, synchronous, waitlist, verbose)
+        job_sheet = self._launch_jobs(jobs, waitlist, verbose)
```
Comment (Zach-Attach) [inline src/nett/nett.py]: Synchronous needs to stay here in order for it work

File: src/nett/nett.py [hunk 10]
```diff
@@ -66,7 +67,6 @@ def run(self,
             steps_per_episode: int = 1000,
             conditions: Optional[list[str]] = None,
             verbose: int = True,
-            synchronous: bool = False,
```
Comment (Zach-Attach) [inline src/nett/nett.py]: Synchronous needs to stay in order to have synchronous option work

File: src/nett/nett.py (hunk 3 above)
Comment (desaibhargav) [reply src/nett/nett.py]: Added

**Review by Zach-Attach — APPROVED** [review]

**Pull request merged**